[Merged by Bors] - feat(topology/homotopy): define nth homotopy group πₙ (without the group instance)

ralvrz opened the pull request:
This pull request adds:
- definition of the nth homotopy group `π n x`
- proof of `π 0 x ≃ zeroth_homotopy X` where `x:X`
- proof of  `π 1 x ≃ path.homotopic.quotient x x` 

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ralvrz committed:
feat(topology/homotopy): define nth homotopy group πₙ, show equivalences of π₀ and π₁

ralvrz committed:
fix line too long

ralvrz committed:
add missing docstrings

ralvrz committed:
fix curly braces lint

ralvrz committed:
fix broken proof

ralvrz committed:
add missing instances

ralvrz committed:
fix

ralvrz committed:
add missing docstring

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+def cube (n:nat) := fin n -> I
+notation `I^` := cube
+
+namespace cube
```
There's some really odd syntax here. 
- We don't indent things inside a namespace. 
- We try to keep `:=` on the same line as the last hypothesis of the theorem statement. 
- We don't indent the beginning of a proof. 
- We try to have spaces after any of `λ`, `,`, `:`. 
- For very small blocks `begin tactic1, tactic2 end`, we prefer `by { tactic1, tactic 2 }`. For bigger blocks, we put `begin` and `end` in their own lines, and indent the contents by two spaces. 
- There should always be an empty line between theorems, unless they're really short and really related.

There's a bunch of golfing opportunities throughout, I'll suggest them after this first cleanup.

ralvrz replied:
Thanks for the suggestions! I will be going through them in the coming days

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+    /--
+    Reversing of a homotopy
+    -/
+    @[symm] def symm : f ~ g -> g ~ f :=
```
Please use `→` instead of `->` throughout.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+end cube
+
+/--
+The continous fuctions I^n->X that send the boundary to x
```
```suggestion
The continuous functions `I^n → X` that send the boundary to `x`.
```

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+  -/
+  def homotopic (f0 f1 : pre_π n x) : Prop := nonempty (f0 ~ f1)
+  namespace homotopic
+    @[refl] lemma refl (f : pre_π n x) : homotopic f f :=
```
Consider also defining an `is_equiv` instance on this relation.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+The nth homotopy group at x defined as the quotient of base preserving functions up to homotopic
+equivalence
+-/
+def π (n : nat) (x : X) := quotient (pre_π.homotopic.setoid n x)
```
This is way too generic of a name to be in the root namespace. Would `homotopy_group` be any better?

Consider renaming `pre_π` as well.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+/--
+The quotient of X up to path components
+-/
+def path_components := quotient (⟨λ x y, x ∈ path_component y,
```
Please add type annotations on all definitions and theorems.

alreadydone replied:
This definition is the same as [zeroth_homotopy](https://leanprover-community.github.io/mathlib_docs/topology/path_connected.html#zeroth_homotopy) so should be dropped.

To construct `pi0_equiv_path_components` you just need to construct an equivalence between `pre_π 0 x` and `X`, show that the `homotopic_rel` on the former corresponds to the `joined` relation on the latter, and use [quotient.congr](https://leanprover-community.github.io/mathlib_docs/logic/equiv/basic.html#quotient.congr).

Similarly, to construct `pi1_equiv_fundamental_group` you just need to construct an equivalence between `pre_π 1 x` and `path x x` respecting homotopy. [path.homotopy](https://leanprover-community.github.io/mathlib_docs/topology/homotopy/path.html#path.homotopy) is already defined in terms of `homotopy_rel`, and the only difference is that `path x x` uses `unit_interval`, aka `I`, and `pre_π` uses `I^1`, i.e. `fin 1 -> I`, so it should be enough to use the homeomorphism between `I` and `fin 1 -> I`, but mathlib doesn't seem to have the composition of a homotopy with a continuous map yet.

vihdzp reviewed (commented):
I know little about homotopies, but there's a lot of style suggestions I can make.

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+  Homotopies are functions I×I^n → X
+  -/
+  def homotopy (f0 f1 : pre_π n x) :=
+    continuous_map.homotopy_rel f0.to_continuous_map f1.to_continuous_map (cube.boundary n)
```
```suggestion
    f0.to_continuous_map.homotopy_rel f1.to_continuous_map (cube.boundary n)
```
should work, but this def seems unnecessary, see below.

ralvrz replied:
Indeed I don't give the group instance as it's still a WIP at the moment.
Thanks for your suggestion, I'll see if in fact makes the work easier

alreadydone replied:
If you do intend to include the group instance in this PR, then it's appropriate to mark the PR with the WIP label. Also I changed the awaiting-review label to awaiting-author; please change it back once you've addressed the comments.

ralvrz replied:
I'm working up to proving commutativity for n>1, I think it'd be nicer to have the abelian group instance its own PR

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+  /--
+  Homotopic relation between `pre_π`s
+  -/
+  def homotopic (f0 f1 : pre_π n x) : Prop := nonempty (f0 ~ f1)
```
Is it necessary to define `homotopy`? Can't you just use [continuous_map.homotopic_rel](https://leanprover-community.github.io/mathlib_docs/topology/homotopy/basic.html#continuous_map.homotopic_rel) here?

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,364 @@
+    := begin ext1 i, rcases i with ⟨⟨_,_⟩,_⟩, reflexivity,
+      exfalso, exact (nat.not_lt_zero _ (nat.le_of_succ_le_succ i_property))
+    end
+end cube
```
```
import logic.unique
lemma one_indep {α} (f : fin 1 → α) (i j : fin 1) : f i = f j :=
by rw subsingleton.elim i j
lemma one_char {α} (x : fin 1 → α) : x = λ _, x 0 :=
by { ext, apply one_indep }
```

alreadydone reviewed (commented):
This PR doesn't show πₙ is a **group** when n is nonzero, and in particular `pi1_equiv_fundamental_group` is not made a group isomorphism; I think this should be made clear in the title / PR description.

Even though the group multiplication (or addition) and negation are not defined, this is indeed the definition that makes the definition of those operations easiest, AFAIK.

ralvrz committed:
fix style

ralvrz committed:
rewrite proofs

ralvrz committed:
renamings and congruence proofs

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -8,22 +8,25 @@ import topology.homotopy.basic
-* `pre_π.homotopy` is the type of homotopies between such functions
-* `π n x` the quotient of `pre_π n x` up to homotopic equivalence
+* `nloop n x` is the type of continous fuctions `I^n→X` that send the boundary to `x`
+* `nth_homotopy_group n x` denoted `π n x` is the quotient of `nloop n x` up to homotopic
```
We don't usually include variable names in declaration names; `homotopy_group n x` should be good enough, but I'm not so sure about `loop` in root namespace.

ralvrz replied:
How about `gen_loop` for generalized loop?

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -66,220 +69,129 @@ The first projection of a non-zero dimensional cube
+
 /--
-The last `n` projection of an `n+1` dimensional cube
+The last `n` projections of an `n+1` dimensional cube
```
This sounds better:

The projection to the last `n` coordinates from an `n+1` dimensional cube

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -66,220 +69,129 @@ The first projection of a non-zero dimensional cube
-  reflexivity,
-end
+def homotopic (f g : nloop n x) : Prop :=
+  nonempty (f.to_continuous_map.homotopy_rel g.to_continuous_map (cube.boundary n))
```
Why not just `f.to_continuous_map.homotopic_rel g.to_continuous_map (cube.boundary n)`? It's defeq to this, and there are API for `homotopic_rel` parallel to `homotopy_rel`.

Maybe you just haven't finished the refactoring.

ralvrz committed:
add missing type annotations

ralvrz committed:
renamings

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+notation `I^` := cube
+
+namespace cube
+instance topology {n} : topological_space (I^n) :=
```
It does not need to be given a name (auto-generated name would be `cube.topological_space` I guess).

And in fact you can just `@[derive [has_zero, has_one, topological_space]]` and remove this. (I did it in my branch.)

alreadydone commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+/-
```
The file should probably be renamed homotopy_group.lean

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+instance unique_cube0 : unique (I^0) :=
+  { default := 0, uniq := by {intro a, ext1, exact x.elim0} }
+
+lemma one_indep {α} (f : fin 1 → α) (i j : fin 1) : f i = f j :=
```
This can be generalized to a `unique` type, and should probably go on a much earlier file.

alreadydone replied:
or just `subsingleton`

alreadydone replied:
I think in this file it's best to state results about `I^1` instead of `fin 1 → α`. I used `fin 1 → α` just to show a minimal working example (MWE) with all necessary imports.
```
lemma one_indep (f : I^1) (i j : fin 1) : f i = f j := by rw subsingleton.elim i j
lemma one_char (x : I^1) : x = λ _, x 0 := by { ext1, apply one_indep }
```
The first one is only used to prove the second one so I think it can be inlined; the second one is used multiple times so it's best to keep it.

I can add the more general results in another PR, but the proofs above are already simple enough so there isn't a clear benefit of using the general versions to prove them.

The following could go after `pi.unique_of_is_empty` in logic/unique:
```
lemma congr_arg_heq_of_subsingleton [subsingleton α] {β : α → Sort v}
  (f : Π a, β a) (i j : α) : f i == f j := by rw subsingleton.elim i j

lemma congr_arg_of_subsingleton [subsingleton α]
  (f : α → β) (i j : α) : f i = f j := by rw subsingleton.elim i j

lemma eq_const_of_unique [unique α] (f : α → β) : f = λ _, f default :=
by { ext, apply congr_arg_of_subsingleton }

lemma heq_const_of_unique [unique α] {β : α → Sort v}
  (f : Π a, β a) : f == λ a : α, f default :=
function.hfunext rfl (λ i _ _, by rw subsingleton.elim i default)
```
but I don't know how useful the heq versions will be.

vihdzp replied:
I like the `heq` versions, even if they're not super useful.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
+
+instance unique_cube0 : unique (I^0) :=
+  { default := 0, uniq := by {intro a, ext1, exact x.elim0} }
```
This shouldn't be indented.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+  { default := 0, uniq := by {intro a, ext1, exact x.elim0} }
+
+lemma one_indep {α} (f : fin 1 → α) (i j : fin 1) : f i = f j :=
+  by rw subsingleton.elim i j
```
This and other `by ...` blocks shouldn't be indented either.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+lemma one_indep {α} (f : fin 1 → α) (i j : fin 1) : f i = f j :=
+  by rw subsingleton.elim i j
+
+lemma one_char {α} (x : fin 1 → α) : x = λ _, x 0 :=
```
(Untested) this is just `unique.eq_default`, since the RHS is def-eq to the default of the type `fin 1 → α`.

alreadydone replied:
but `unique (fin 1 → α)` isn't true.

vihdzp replied:
My bad.

vihdzp commented on src/topology/homotopy/pi_n.lean:
```diff
@@ -0,0 +1,240 @@
+The nth dimensional generalized loops; functions `I^n → X` fixed at the boundary
+-/
+structure gen_loop (n : nat) (x : X) extends C(I^n,X) :=
+  (boundary : ∀ y ∈ cube.boundary n, to_fun y = x )
```
This shouldn't be indented either.

vihdzp reviewed (commented):
A few more style remarks.

ralvrz committed:
remove indentations

alreadydone committed:
golf + docstring

alreadydone reviewed (commented):
The PR looks mostly ready now!
I golfed some proofs and made some docstring/formatting improvements at [branch pi_n_golfs](https://github.com/leanprover-community/mathlib/compare/pi_n...pi_n_golfs?expand=1); please adopt what you agree with and let me know where you have different opinions.

alreadydone committed:
slight golf

ralvrz committed:
Merge branch 'pi_n_golfs' into pi_n

ralvrz committed:
add missing docstring

ralvrz committed:
rename file

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,209 @@
+import topology.homotopy.basic
+import topology.path_connected
+import topology.unit_interval
+import algebraic_topology.fundamental_groupoid.fundamental_group
```
This is the only import necessary, the others are transitively imported.

ralvrz replied:
Defining concatenation by coordinate is turning out to be troublesome, continuity in particular, also I haven't been able to define it in terms of path concatenation so that the code in [fundamental_groupoid/basic](https://github.com/leanprover-community/mathlib/blob/master/src/algebraic_topology/fundamental_groupoid/basic.lean) can be reused, so there's much code duplication.

I think using the currying homeomorphism is a better idea, thanks for the pointer!

alreadydone replied:
If we adopt the currying approach, I think the first step is to construct `gen_loop (n+1) x ≃ path const const`, where `const : gen_loop n x`, with the compact-open topology on `gen_loop n x` as a subspace of `C(I^n, X)`. (You get one such equivalence for each of the (n+1) coordinates.) Then, we would show that the type [path.homotopy](https://leanprover-community.github.io/mathlib_docs/topology/homotopy/path.html#path.homotopy) between two `path const const` is in bijection with `homotopy_rel` boundary between the two corresponding `gen_loop (n+1) x`. That would allow us to take full advantage of the existing fundamental groupoid constructions and theorems.

In fact, I suspect the constructions will be very similar to what we've already done in `pi1_equiv_fundamental_group` ...

alreadydone replied:
Iterating `gen_loop (n+1) x ≃ path const const` + some continuity argument should allow us to show that `gen_loop n x` is homeomorphic to the `n`th iterated loop space `X`, so I would not object if you change the name to `loop_space` :) (Notation like `Ω^[n] X` would also be nice.)

ralvrz replied:
The iterated loop space homeomorphism was also suggested by Prof. Bartholdi, however unless it makes an earlier goal easier it'll remain later in the queue.

The priority goes like this: group/abelian instace > paths induce homomorphisms functorially > group homomorphisms instances > iterated loop space homeomorphism

urkud commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,209 @@
+  apply continuous_infi_dom,
+  rw continuous_def,
+  exact λ s H, is_open_induced_iff'.2 ⟨s, H, rfl⟩
+end
```
```suggestion
continuous_apply i
```

ralvrz replied:
Excellent! That proof is essentially refl but I didn't knew how to golf it

urkud commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,209 @@
+
+namespace cube
+
+@[continuity] lemma proj_continuous {n} (i : fin n) : continuous (λ f : I^n, f i) :=
```
Here and below: you already have `variable {n}`, no need to repeat it.
```suggestion
@[continuity] lemma proj_continuous (i : fin n) : continuous (λ f : I^n, f i) :=
```

urkud commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,209 @@
+@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
+
+instance unique_cube0 : unique (I^0) :=
+{ default := 0, uniq := by {intro a, ext1, exact x.elim0} }
```
`pi.unique_of_is_empty`?

urkud commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,209 @@
+The `n`th homotopy group at `x` defined as the quotient of `gen_loop n x` by the
+`homotopic` relation.
+-/
+def homotopy_group (n : ℕ) (x : X) : Type _ := quotient (gen_loop.homotopic.setoid n x)
```
```suggestion
@[derive inhabited]
def homotopy_group (n : ℕ) (x : X) : Type _ := quotient (gen_loop.homotopic.setoid n x)
```

urkud commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,209 @@
+
+instance homotopy_group.inhabited : inhabited (π n x) :=
+{ default :=  quotient.mk' gen_loop.const }
+
```
```suggestion
```

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

ralvrz committed:
unique instance of I^0

alreadydone reviewed (approved):
LGTM now, but I suggest you add `(without the group structure yet)` in the PR title.

As for the next steps, I think it's easiest to use [eckmann_hilton.comm_group](https://leanprover-community.github.io/mathlib_docs/group_theory/eckmann_hilton.html#eckmann_hilton.comm_group) to show the abelian-ness when n > 1. We should probably use the "concatenation" along the first axis (`head`) as the default group operation `*`, and `m₁` for the same in the second axis. We more or less just need to copy the development in [fundamental_groupoid/basic](https://github.com/leanprover-community/mathlib/blob/master/src/algebraic_topology/fundamental_groupoid/basic.lean), just some path reparametrizations, to show that `π n x` is a group under both `*` and `m₁`; the `distrib` condition should hold even before the quotient.

We should probably factor out the path reparametrization code to reduce code duplication, or alternatively, we could take advantage of `C(I^(n+1),X) ≃ C(I,C(I^n,X))` with the compact-open topology on `C(I^n,X)` ([this result](https://ncatlab.org/nlab/show/exponential+law+for+spaces#general_exponential_laws) seems [already in mathlib](https://leanprover-community.github.io/mathlib_docs/topology/compact_open.html)) and use the results about fundamental groupoid directly.

ralvrz committed:
remove one_indep, change fin 1 → α with I^1

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,199 @@
+instance unique_cube0 : unique (I^0) := pi.unique_of_is_empty _
+
+lemma one_char (f : I^1) : f = λ _, f 0 :=
+by { ext, rwa (subsingleton.elim x 0) }
```
```suggestion
by { ext, rwa subsingleton.elim x 0 }
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,199 @@
+import topology.homotopy.basic
+import topology.path_connected
+import topology.unit_interval
+import algebraic_topology.fundamental_groupoid.fundamental_group
```
```suggestion
import algebraic_topology.fundamental_groupoid.fundamental_group
```

alreadydone reviewed (commented):
Last night I started a discussion about the notation (+ or *) for the group operation on homotopy groups at the [homotopy formalization Zulip topic](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Homotopy.20formalization/near/286630228) you opened last year. You didn't reply so I assume you didn't notice. From the opinions gathered, [the best solution](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Homotopy.20formalization/near/286653843) seems to be to define a `group` instance on pi_{n+1} (yielding * notation) and a `add_comm_group` instance on pi_{n+2} (yielding + notation) with the same group structure. And we could provide a version with pi_n but with the conditions `[fact (0 < n)]` and `[fact (1 < n)]` respectively. 

About the notation of the type itself, it seems we should definitely make X explicit. And [this](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Homotopy.20formalization/near/286674088) is so far the best suggestion, allowing both notation of the form `π_ n ⟨X⟩` and of the form `π_ n ⟨X, x⟩`. I'm not sure this notation should be localized or global.

I think this PR could be merged as is (without group instances) but maintainers may have different opinions about what [minimal test](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Homotopy.20formalization/near/286649216) you should prove to ensure the definition works, and you may ask them in that topic.

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

alreadydone committed:
feat(logic/unique): functions from a `unique` type is `const` (#14823)

+ A function `f` from a `unique` type is equal to the constant function with value `f default`, and the analogous heq version for dependent functions.

+ Also changes `Π a : α, Sort v` in the file to `α → Sort v`.

Inspired by https://github.com/leanprover-community/mathlib/pull/14724#discussion_r900542203

ralvrz committed:
one_char by eq_const_of_unique

kbuzzard commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -0,0 +1,195 @@
+-/
+@[derive [has_zero, has_one, topological_space]]
+def cube (n : ℕ) : Type := fin n → I
+notation `I^` := cube
```
Should this be local notation for this file?

urkud replied:
Or localized in `unit_interval`?

ralvrz replied:
I've changed it to a local notation for now.

kbuzzard reviewed (approved):
This looks very nice to me -- thanks a lot for your contribution!

ralvrz committed:
local `I^` notation
